var/spack/repos/builtin/packages/pumi/package.py

```diff
@@ -108,22 +108,28 @@ def cmake_args(self):
     def test(self):
         if self.spec.version <= Version("2.2.6"):
             return
-        exe = "uniform"
-        options = ["../testdata/pipe.dmg", "../testdata/pipe.smb", "pipe_unif.smb"]
-        expected = "mesh pipe_unif.smb written"
+        data_dir = self.prefix.share.testdata
+        exe = self.prefix.bin.uniform
+        options = [
+            join_path(data_dir, "pipe.dmg"),
+            join_path(data_dir, "pipe.smb"),
+            join_path(self.prefix.bin, "pipe_unif.smb"),
+        ]
+        expected = "pipe_unif.smb written"
         description = "testing pumi uniform mesh refinement"
-        self.run_test(exe, options, expected, purpose=description, work_dir=self.prefix.bin)
+        self.run_test(exe, options, expected, purpose=description)
 
-        mpiexec = Executable(join_path(self.spec["mpi"].prefix.bin, "mpiexec")).command
-        mpiopt = ["-n", "2"]
-        exe = ["split"]
-        options = ["../testdata/pipe.dmg", "../testdata/pipe.smb", "pipe_2_.smb", "2"]
-        expected = "mesh pipe_2_.smb written"
+        mpiexec = self.spec["mpi"].prefix.bin.mpiexec
```

Comment:
Is there a case where the full path to the mpi executable is necessary? There are software environments where only the mpi compilers are provided in the mpi bin directory and the executors are located elsewhere. This test is breaking for us in that situation. Also, mpiexec is not always available. I recommend using something like this:

```
        mpiexe_list = ["mpirun", "mpiexec", "srun"]
        for mpiexe in mpiexe_list:
            if which(mpiexe) is not None:
                mpiexec = Executable(mpiexe).command
                break
```

Reply:
I'm not sure.  If the full path was not set, how would the spack test mechanism locate the mpi executor that matches the one used for the build?  I haven't tried mixing an mpi executor from one mpi install with libraries/headers from another.

Reply:
The correct mpi environment would have to be loaded for the test to run successfully so it would pick up the executor from the path. I think that is generally how spack smoke tests are implemented. This isn't to support mixing and matching mpi components but a single mpi with an atypical directory structure.

Reply:
Sounds good.  Thank you.  I'll make the change.

Reply:
Quick spack/python question.  Should we handle the case where `mpirun`, `mpiexec` or `srun` is not found?

Reply:
That would be safer. Maybe skip the mpi test with a warning rather than a failure in that case? Ideally we should have an mpi executor method provided by spack so every mpi enabled test doesn't have to reimplement this...